target/linux/ipq807x/base-files/lib/upgrade/platform.sh

```diff
@@ -0,0 +1,34 @@
+PART_NAME=firmware
+REQUIRE_IMAGE_METADATA=1
+
+RAMFS_COPY_BIN='fw_printenv fw_setenv'
+RAMFS_COPY_DATA='/etc/fw_env.config /var/lock/fw_printenv.lock'
+
+platform_check_image() {
+	return 0;
+}
+
+platform_do_upgrade() {
+	case "$(board_name)" in
+	xiaomi,ax3600)
+		# Enforce single partition.
+		fw_setenv flag_boot_rootfs 0
+		fw_setenv flag_last_success 0
+		fw_setenv flag_boot_success 0
+		fw_setenv flag_try_sys1_failed 0
+
+		# Second partition won't work and can't be used
+		# Flag is as failed by default
+		fw_setenv flag_try_sys2_failed 1
+
+		# kernel and rootfs are placed on 2 different ubi
+		# First ubiformat the kernel partition than do nand upgrade
```

Comment:
It doesn't do that. nand_do_upgrade only reformat the partitions if it fails to attach it, and it may successfully attach a UBI partition from leftover data even if the partition offset/size changes. Maybe explicitly format the two partitions if the system is started from initramfs?

Reply:
This is correct, I was relying on the fact that UBI attach on rootfs will always fail as multiple partitions have been merged into it.

The comment itself is incorrect and outdated as the kernel should already be ready since from the stock FW you already ubiformated it with an initramfs kernel so it can easily be updated.

But I see your point, it would be beneficial to ubiformat them both if running from initramfs, I am gonna add that.

Reply:
Actually, it looks like check for initramfs is completely broken and will always be true so ubiformat will get triggered every time

Reply:
platform_do_upgrade is executed after switching to a ramfs root. Have you tried checking it in platform_pre_upgrade, like what's done for Redmi AX6000 in mediatek?

Reply:
> Actually, it looks like check for initramfs is completely broken and will always be true so ubiformat will get triggered every time

this looks interesting?

Reply:
Like @981213 mentioned, its because `platform_do_upgrade` is invoked after we already switched to ramdisk.

But, I will try using `platform_pre_upgrade` for it

Reply:
Should be sorted out with the latest push, thanks